addOns/ratelimit/src/main/java/org/zaproxy/addon/ratelimit/OptionsRateLimitPanel.java

```diff
@@ -0,0 +1,199 @@
+
+        this.add(new JLabel(Constant.messages.getString("ratelimit.options.label.tokens")), gbc);
+
+        rateLimitOptionsPanel = new RateLimitMultipleOptionsPanel(getRateLimitTableModel());
+
+        gbc.weighty = 1.0;
+        this.add(rateLimitOptionsPanel, gbc);
+    }
+
+    @Override
+    public void initParam(Object obj) {
+        OptionsParam optionsParam = (OptionsParam) obj;
+        RateLimitParam param = optionsParam.getParamSet(RateLimitParam.class);
+        getRateLimitTableModel().setRules(param.getRules());
+        rateLimitOptionsPanel.setRateLimitParam(param);
+    }
+
+    @Override
+    public void validateParam(Object obj) throws Exception {}
+
+    @Override
+    public void saveParam(Object obj) throws Exception {
+        OptionsParam optionsParam = (OptionsParam) obj;
+        RateLimitParam replacerParam = optionsParam.getParamSet(RateLimitParam.class);
+        replacerParam.setRules(getRateLimitTableModel().getElements());
+    }
+
+    private OptionsRateLimitTableModel getRateLimitTableModel() {
+        if (rateLimitModel == null) {
+            rateLimitModel = new OptionsRateLimitTableModel();
+        }
+        return rateLimitModel;
+    }
+
+    @Override
+    public String getHelpIndex() {
+        return "ratelimit";
+    }
+
+    private static class RateLimitMultipleOptionsPanel
+            extends AbstractMultipleOptionsTablePanel<RateLimitParamRule> {
+
+        private static final long serialVersionUID = -115340627058929308L;
+
+        private static final String REMOVE_DIALOG_TITLE =
+                Constant.messages.getString("ratelimit.options.dialog.token.remove.title");
+        private static final String REMOVE_DIALOG_TEXT =
+                Constant.messages.getString("ratelimit.options.dialog.token.remove.text");
+
+        private static final String REMOVE_DIALOG_CONFIRM_BUTTON_LABEL =
+                Constant.messages.getString("ratelimit.options.dialog.token.remove.button.confirm");
+        private static final String REMOVE_DIALOG_CANCEL_BUTTON_LABEL =
+                Constant.messages.getString("ratelimit.options.dialog.token.remove.button.cancel");
+
+        private static final String REMOVE_DIALOG_CHECKBOX_LABEL =
+                Constant.messages.getString("ratelimit.options.dialog.token.remove.checkbox.label");
+
+        private RateLimitParam rateLimitParam;
+        private RateLimitRuleAddDialog addDialog;
+        private RateLimitRuleModifyDialog modifyDialog;
+
+        public RateLimitMultipleOptionsPanel(OptionsRateLimitTableModel model) {
+            super(model);
+
+            this.model = model;
+
+            getTable().getColumnExt(0).setPreferredWidth(25); // checkbox column should be tiny
+            getTable().getColumnExt(1).setPreferredWidth(200); // description
+            getTable().getColumnExt(2).setPreferredWidth(100); // matchString
+            getTable().getColumnExt(3).setPreferredWidth(50); // requestsPerSecond
```

Comment:
As above, if possible it would be nice to have this be fully dynamic.

Reply:
fixed